File: src/mpc_controller.py
"""
MPC Controller Module - Model Predictive Control for Lane Keeping
Part of the 3D Robotics Lab simulation.
"""

import pygame
from pygame.locals import *
from OpenGL.GL import *
from OpenGL.GLU import *
import numpy as np


class MPCLaneKeeping:
    """
    Model Predictive Control Lane Keeping Assist

    Uses forward simulation with the Ackermann kinematic model to predict
    future trajectories for different steering commands, then selects the
    command that minimizes a cost function.

    Cost function penalizes:
    - Lateral deviation from lane center
    - Leaving the road (off-track)
    - Crossing into opposite lane
    """

    def __init__(self, car, camera):
        self.car = car
        self.camera = camera
        self.active = False
        self.was_manually_overridden = False

        # MPC parameters - TUNED for smoother behavior
        self.prediction_horizon = 20  # Look further ahead for smoother planning
        self.dt = 0.15  # Larger time steps = less reactive
        self.num_candidates = 9  # Fewer candidates = faster, less twitchy

        # Cost function weights - SOFTENED for less aggressive behavior
        self.weight_lateral_deviation = 20.0  # Reduced from 100 - less aggressive corrections
        self.weight_off_track = 10000.0  # Keep strong - safety critical
        self.weight_wrong_lane = 5000.0  # Keep strong - safety critical
        self.weight_steering_effort = 10.0  # Increased from 1 - prefer less steering
        self.weight_steering_change = 50.0  # NEW - penalize rapid steering changes
        self.weight_boundary_proximity = 200.0  # NEW - heavily penalize getting close to boundaries

        # Steering command candidates (fraction of max steering) - MORE CONSERVATIVE
        self.steering_fractions = np.linspace(-0.6, 0.6, self.num_candidates)  # Limited to ±60% instead of ±100%

        # Store predicted trajectory for visualization
        self.predicted_trajectory = []  # List of (x, y) positions
        self.selected_steering = 0.0
        self.best_cost = float('inf')
        self.previous_steering = 0.0  # Track previous steering for smoothness

    def toggle(self):
        """Toggle MPC LKA on/off"""
        self.active = not self.active
        self.was_manually_overridden = False
        return self.active

    def deactivate(self):
        """Deactivate MPC LKA"""
        if self.active:
            self.active = False
            self.was_manually_overridden = True

    def calculate_steering(self, track):
        """
        MPC control loop:
        1. Generate candidate steering commands
        2. Simulate forward for each candidate
        3. Compute cost for each trajectory
        4. Select command with minimum cost
        """
        if not self.active:
            self.predicted_trajectory = []
            return None

        # Get current lane detection
        left_lane, right_lane, center_lane = self.camera.detect_lanes(track)

        # Determine which lane we're in
        current_lane = self.camera.current_lane

        if current_lane == "LEFT":
            lane_left_boundary = left_lane
            lane_right_boundary = center_lane
        elif current_lane == "RIGHT":
            lane_left_boundary = center_lane
            lane_right_boundary = right_lane
        else:
            # Unknown lane - can't compute costs properly
            self.predicted_trajectory = []
            return None

        # Need at least some lane points for cost calculation
        if len(lane_left_boundary) == 0 or len(lane_right_boundary) == 0:
            self.predicted_trajectory = []
            return None

        # Compute lane center points for cost function
        lane_center_points = self._compute_lane_centers(
            lane_left_boundary, lane_right_boundary
        )

        if len(lane_center_points) == 0:
            self.predicted_trajectory = []
            return None

        # Current state
        current_x = self.car.x
        current_y = self.car.y
        current_theta = self.car.theta
        current_velocity = self.car.velocity

        # Evaluate all candidate steering commands
        best_steering = 0.0
        best_cost = float('inf')
        best_trajectory = []

        for steering_fraction in self.steering_fractions:
            # Convert fraction to actual steering angle
            steering_cmd = steering_fraction * self.car.max_steering_angle

            # Simulate forward with this steering command
            trajectory, cost = self._simulate_and_cost(
                current_x, current_y, current_theta, current_velocity,
                steering_cmd, lane_center_points, track, current_lane
            )

            # Track best command
            if cost < best_cost:
                best_cost = cost
                best_steering = steering_cmd
                best_trajectory = trajectory

        # Store for visualization
        self.predicted_trajectory = best_trajectory
        self.selected_steering = best_steering
        self.best_cost = best_cost

        # Update previous steering for next iteration (smoother transitions)
        self.previous_steering = best_steering

        return best_steering

    def _compute_lane_centers(self, left_boundary, right_boundary):
        """Compute lane center points from left and right boundaries"""
        lane_centers = []

        # Pair closest points
        for left_x, left_y, _ in left_boundary:
            min_dist = float('inf')
            closest_right = None

            for right_x, right_y, _ in right_boundary:
                dist = np.sqrt((right_x - left_x)**2 + (right_y - left_y)**2)
                if dist < min_dist:
                    min_dist = dist
                    closest_right = (right_x, right_y)

            if closest_right:
                center_x = (left_x + closest_right[0]) / 2
                center_y = (left_y + closest_right[1]) / 2
                lane_centers.append((center_x, center_y))

        return lane_centers

    def _simulate_and_cost(self, start_x, start_y, start_theta, start_velocity,
                          steering_cmd, lane_centers, track, current_lane):
        """
        Simulate vehicle forward with constant steering command
        and compute trajectory cost
        """
        trajectory = []
        total_cost = 0.0

        # Simulate state
        x = start_x
        y = start_y
        theta = start_theta
        velocity = start_velocity

        wheelbase = self.car.wheelbase

        for step in range(self.prediction_horizon):
            # Store position
            trajectory.append((x, y))

            # Ackermann kinematic model update
            if abs(velocity) > 0.1:
                omega = velocity * np.tan(steering_cmd) / wheelbase
                x += velocity * np.cos(theta) * self.dt
                y += velocity * np.sin(theta) * self.dt
                theta += omega * self.dt
                theta = np.arctan2(np.sin(theta), np.cos(theta))

            # Compute cost at this position
            step_cost = self._compute_position_cost(
                x, y, lane_centers, track, current_lane
            )

            total_cost += step_cost

            # Early termination if cost is already too high
            if total_cost > 1e6:
                break

        # Add steering effort cost (penalize large steering angles)
        total_cost += self.weight_steering_effort * abs(steering_cmd)

        # Add steering change cost (penalize rapid changes from previous steering)
        steering_change = abs(steering_cmd - self.previous_steering)
        total_cost += self.weight_steering_change * steering_change

        return trajectory, total_cost

    def _compute_position_cost(self, x, y, lane_centers, track, current_lane):
        """
        Compute cost for a given position:
        - Lateral deviation from lane center
        - Penalty for being off-track
        - Penalty for being in wrong lane
        """
        cost = 0.0

        # 1. Lateral deviation from lane center
        if len(lane_centers) > 0:
            # Find closest lane center point
            distances = [np.sqrt((x - cx)**2 + (y - cy)**2)
                        for cx, cy in lane_centers]
            min_dist = min(distances)
            closest_idx = distances.index(min_dist)
            closest_center = lane_centers[closest_idx]

            # Lateral deviation (linear cost for smoother behavior near center)
            lateral_dev = np.sqrt((x - closest_center[0])**2 +
                                 (y - closest_center[1])**2)
            cost += self.weight_lateral_deviation * lateral_dev

        # 2. Check if position is on track
        on_track, lateral_offset = self._is_on_track(x, y, track)

        if not on_track:
            # Strongly penalize being off the road
            cost += self.weight_off_track
        else:
            # 3. Boundary proximity cost - exponentially increase as we approach edges
            # Track has total width = 2 * lane_width, so half_width = track_width / 2
            half_width = track.track_width / 2.0
            distance_to_boundary = half_width - abs(lateral_offset)

            # Exponential cost that increases rapidly as we approach the boundary
            # When distance_to_boundary is small, cost is high
            if distance_to_boundary > 0:
                # Normalize: 0 = at boundary (edge), 1 = at centerline
                normalized_distance = distance_to_boundary / half_width
                # Exponential penalty: cost increases as normalized_distance decreases
                # e^(-3*x) gives strong penalty when x is small (near boundary)
                proximity_cost = np.exp(-3.0 * normalized_distance)
                cost += self.weight_boundary_proximity * proximity_cost

            # 4. Check if in wrong lane
            in_wrong_lane = self._is_in_wrong_lane(lateral_offset, current_lane, track.lane_width)
            if in_wrong_lane:
                cost += self.weight_wrong_lane

        return cost

    def _is_on_track(self, x, y, track):
        """
        Check if position (x, y) is on the track
        Returns: (on_track: bool, lateral_offset: float)
        """
        # Find closest centerline point
        min_dist = float('inf')
        closest_idx = 0

        for i, (cx, cy) in enumerate(track.centerline):
            dist = np.sqrt((x - cx)**2 + (y - cy)**2)
            if dist < min_dist:
                min_dist = dist
                closest_idx = i

        # Compute lateral offset from centerline
        p_curr = track.centerline[closest_idx]
        p_next = track.centerline[(closest_idx + 1) % len(track.centerline)]

        dx = p_next[0] - p_curr[0]
        dy = p_next[1] - p_curr[1]
        track_angle = np.arctan2(dy, dx)

        to_point_x = x - p_curr[0]
        to_point_y = y - p_curr[1]

        perp_angle = track_angle + np.pi / 2
        lateral_offset = (to_point_x * np.cos(perp_angle) +
                         to_point_y * np.sin(perp_angle))

        # Check if within track bounds
        # Track width is total width (2 * lane_width), so edges are at ±track_width/2
        half_width = track.track_width / 2.0
        on_track = abs(lateral_offset) <= half_width

        return on_track, lateral_offset

    def _is_in_wrong_lane(self, lateral_offset, current_lane, lane_width):
        """
        Check if lateral offset corresponds to being in the wrong lane
        
        Track geometry:
        - Track centerline is at lateral_offset = 0
        - Left lane center is at -lane_width/2
        - Right lane center is at +lane_width/2
        - Lane boundary (center line) is at lateral_offset = 0
        """
        half_lane = lane_width / 2.0
        
        if current_lane == "LEFT":
            # Left lane center is at -half_lane
            # Wrong if we cross centerline into right lane (positive offset > half_lane)
            return lateral_offset > half_lane
        elif current_lane == "RIGHT":
            # Right lane center is at +half_lane
            # Wrong if we cross centerline into left lane (negative offset < -half_lane)
            return lateral_offset < -half_lane
        else:
            return False

File: cpp/src/mpc_controller.cpp
#include "mpc_controller.h"
#include "utils.h"
#include <omp.h>
#include <cmath>
#include <algorithm>
#include <limits>

namespace lka {

MPCController::MPCController(const MPCConfig& config)
    : config_(config) {
}

VehicleState MPCController::forward_simulate(const VehicleState& state,
                                             double steering_angle,
                                             double dt) const {
    VehicleState next = state;

    // Ackermann steering model (kinematic bicycle model)
    // Clamp steering angle
    double steering = math::clamp(steering_angle,
                                  -config_.max_steering_angle,
                                  config_.max_steering_angle);

    // Update yaw rate
    double yaw_rate = (state.speed / config_.wheelbase) * std::tan(steering);

    // Update position and yaw
    next.yaw = state.yaw + yaw_rate * dt;
    next.yaw = math::normalize_angle(next.yaw);

    next.x = state.x + state.speed * std::cos(next.yaw) * dt;
    next.y = state.y + state.speed * std::sin(next.yaw) * dt;

    // Speed remains constant in this simplified model
    next.speed = state.speed;

    return next;
}

void MPCController::get_lane_reference(double x, double y,
                                       const LaneBoundary& left_boundary,
                                       const LaneBoundary& right_boundary,
                                       double& lane_center_y,
                                       double& lane_heading) const {
    // Find closest points on both boundaries
    double min_dist_left = std::numeric_limits<double>::max();
    double min_dist_right = std::numeric_limits<double>::max();
    size_t idx_left = 0;
    size_t idx_right = 0;

    // Find closest left boundary point
    for (size_t i = 0; i < left_boundary.points.size(); ++i) {
        double dist = math::distance_squared(x, y,
                                             left_boundary.points[i].x,
                                             left_boundary.points[i].y);
        if (dist < min_dist_left) {
            min_dist_left = dist;
            idx_left = i;
        }
    }

    // Find closest right boundary point
    for (size_t i = 0; i < right_boundary.points.size(); ++i) {
        double dist = math::distance_squared(x, y,
                                             right_boundary.points[i].x,
                                             right_boundary.points[i].y);
        if (dist < min_dist_right) {
            min_dist_right = dist;
            idx_right = i;
        }
    }

    // Compute lane center
    if (left_boundary.points.empty() || right_boundary.points.empty()) {
        lane_center_y = y;
        lane_heading = 0;
        return;
    }

    const Point3D& left_pt = left_boundary.points[idx_left];
    const Point3D& right_pt = right_boundary.points[idx_right];

    lane_center_y = (left_pt.y + right_pt.y) / 2.0;

    // Estimate lane heading from adjacent points
    if (idx_left < left_boundary.points.size() - 1 && idx_right < right_boundary.points.size() - 1) {
        const Point3D& left_next = left_boundary.points[idx_left + 1];
        const Point3D& right_next = right_boundary.points[idx_right + 1];

        double heading_left = std::atan2(left_next.y - left_pt.y, left_next.x - left_pt.x);
        double heading_right = std::atan2(right_next.y - right_pt.y, right_next.x - right_pt.x);

        lane_heading = (heading_left + heading_right) / 2.0;
    } else {
        lane_heading = 0;
    }
}

double MPCController::distance_to_boundaries(double x, double y,
                                             const LaneBoundary& left_boundary,
                                             const LaneBoundary& right_boundary) const {
    double min_dist_left = std::numeric_limits<double>::max();
    double min_dist_right = std::numeric_limits<double>::max();

    // Find minimum distance to left boundary
    for (const auto& pt : left_boundary.points) {
        double dist = std::abs(y - pt.y); // Simplified lateral distance
        if (dist < min_dist_left) {
            min_dist_left = dist;
        }
    }

    // Find minimum distance to right boundary
    for (const auto& pt : right_boundary.points) {
        double dist = std::abs(y - pt.y); // Simplified lateral distance
        if (dist < min_dist_right) {
            min_dist_right = dist;
        }
    }

    return std::min(min_dist_left, min_dist_right);
}

double MPCController::compute_trajectory_cost(const VehicleState& initial_state,
                                              double steering_angle,
                                              const LaneBoundary& left_boundary,
                                              const LaneBoundary& right_boundary,
                                              double current_steering) const {
    double total_cost = 0.0;
    VehicleState state = initial_state;

    // Simulate forward for prediction horizon
    for (int step = 0; step < config_.prediction_horizon; ++step) {
        // Forward simulate one step
        state = forward_simulate(state, steering_angle, config_.dt);

        // Get lane reference at this position
        double lane_center_y, lane_heading;
        get_lane_reference(state.x, state.y, left_boundary, right_boundary,
                          lane_center_y, lane_heading);

        // Lane center tracking cost
        double lateral_error = state.y - lane_center_y;
        total_cost += config_.weight_lane_center * lateral_error * lateral_error;

        // Heading alignment cost
        double heading_error = math::normalize_angle(state.yaw - lane_heading);
        total_cost += config_.weight_heading * heading_error * heading_error;

        // Boundary proximity penalty (higher cost near boundaries)
        double boundary_dist = distance_to_boundaries(state.x, state.y,
                                                      left_boundary, right_boundary);
        if (boundary_dist < 1.0) {
            total_cost += 100.0 * (1.0 - boundary_dist) * (1.0 - boundary_dist);
        }
    }

    // Steering effort cost
    total_cost += config_.weight_steering_effort * steering_angle * steering_angle;

    // Steering smoothness cost (penalize large changes from current steering)
    double steering_change = steering_angle - current_steering;
    total_cost += config_.weight_steering_smoothness * steering_change * steering_change;

    return total_cost;
}

double MPCController::calculate_steering(const VehicleState& state,
                                         const LaneBoundary& left_boundary,
                                         const LaneBoundary& right_boundary,
                                         double current_steering) {
    // Generate steering candidates
    std::vector<double> candidates(config_.num_candidates);
    for (int i = 0; i < config_.num_candidates; ++i) {
        double t = static_cast<double>(i) / (config_.num_candidates - 1);
        candidates[i] = -config_.max_steering_angle +
                       t * 2.0 * config_.max_steering_angle;
    }

    // Evaluate all candidates in parallel
    std::vector<double> costs(config_.num_candidates);
    std::vector<std::vector<Point2D>> trajectories(config_.num_candidates);

    #pragma omp parallel for schedule(dynamic)
    for (int i = 0; i < config_.num_candidates; ++i) {
        costs[i] = compute_trajectory_cost(state, candidates[i],
                                           left_boundary, right_boundary,
                                           current_steering);

        // Store trajectory for visualization
        VehicleState sim_state = state;
        trajectories[i].reserve(config_.prediction_horizon);
        for (int step = 0; step < config_.prediction_horizon; ++step) {
            sim_state = forward_simulate(sim_state, candidates[i], config_.dt);
            trajectories[i].emplace_back(sim_state.x, sim_state.y);
        }
    }

    // Find best candidate
    int best_idx = 0;
    double best_cost = costs[0];
    for (int i = 1; i < config_.num_candidates; ++i) {
        if (costs[i] < best_cost) {
            best_cost = costs[i];
            best_idx = i;
        }
    }

    // Store predicted trajectory for visualization
    predicted_trajectory_ = trajectories[best_idx];

    return candidates[best_idx];
}

} // namespace lka
